File: ipad-app/EigoLens/Screens/Settings/SettingsScreen.swift
import SwiftUI

struct SettingsScreen: View {
    var onBack: () -> Void
    @EnvironmentObject var container: AppContainer
    @StateObject private var viewModel = SettingsViewModel()

    var body: some View {
        List {
            // Account Section
            Section("Account") {
                if let email = container.authManager.authState.email {
                    HStack {
                        Text("Signed in as")
                        Spacer()
                        Text(email)
                            .foregroundStyle(.secondary)
                    }
                    Button("Sign Out", role: .destructive) {
                        viewModel.signOut()
                    }
                } else {
                    Text("Guest Mode")
                        .foregroundStyle(.secondary)
                }
            }

            // AI Analysis Section
            Section("AI Analysis") {
                // Provider selection
                Picker("Active Provider", selection: $viewModel.activeProvider) {
                    ForEach(["Claude", "Gemini"], id: \.self) { provider in
                        Text(provider).tag(provider as String?)
                    }
                }
                .onChange(of: viewModel.activeProvider) { _, newValue in
                    viewModel.setActiveProvider(newValue)
                }
            }

            // API Keys Section
            Section("API Keys") {
                apiKeyEditor(
                    label: "Claude API Key",
                    key: $viewModel.claudeApiKey,
                    placeholder: "sk-ant-...",
                    onSave: { viewModel.saveClaudeKey() }
                )

                apiKeyEditor(
                    label: "Gemini API Key",
                    key: $viewModel.geminiApiKey,
                    placeholder: "AI...",
                    onSave: { viewModel.saveGeminiKey() }
                )
            }

            // Token Usage Section
            Section("Token Usage") {
                tokenUsageCard(
                    providerName: "Claude Haiku",
                    inputTokens: viewModel.claudeInputTokens,
                    outputTokens: viewModel.claudeOutputTokens,
                    pricing: TokenUsageStore.claudePricing,
                    onReset: { viewModel.resetTokenUsage(provider: "Claude") }
                )
                tokenUsageCard(
                    providerName: "Gemini 2.5 Flash",
                    inputTokens: viewModel.geminiInputTokens,
                    outputTokens: viewModel.geminiOutputTokens,
                    pricing: TokenUsageStore.geminiPricing,
                    onReset: { viewModel.resetTokenUsage(provider: "Gemini") }
                )
            }

            // About Section
            Section("About") {
                infoRow(label: "Version", value: Bundle.main.infoPlistString(for: "CFBundleShortVersionString") ?? "1.0.0")
                infoRow(label: "Build", value: Bundle.main.infoPlistString(for: "CFBundleVersion") ?? "1")
                infoRow(label: "OCR Engine", value: "Apple Vision Framework")
                infoRow(label: "Dictionary", value: "WordNet")
            }

            // Features Section
            Section("Features") {
                featureRow(icon: "camera.fill", title: "Camera OCR", description: "Capture text from photos")
                featureRow(icon: "textformat.abc", title: "Word Definitions", description: "WordNet with NLP lemmatization")
                featureRow(icon: "brain", title: "AI Analysis", description: "Claude & Gemini integration")
                featureRow(icon: "chart.bar.fill", title: "Readability", description: "FK Grade, SMOG, Coleman-Liau, Flesch")
            }
        }
        .navigationTitle("Settings")
        .navigationBarTitleDisplayMode(.large)
        .toolbar {
            ToolbarItem(placement: .navigationBarLeading) {
                Button(action: onBack) {
                    Image(systemName: "chevron.left")
                }
            }
        }
        .task {
            viewModel.configure(container: container)
        }
        .preferredColorScheme(.dark)
    }

    // MARK: - API Key Editor

    private func apiKeyEditor(label: String, key: Binding<String>, placeholder: String, onSave: @escaping () -> Void) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(label)
                .font(EigoLensTheme.titleMedium)
            HStack {
                SecureField(placeholder, text: key)
                    .textFieldStyle(.roundedBorder)
                    .font(EigoLensTheme.bodyMedium)

                Button("Save", action: onSave)
                    .buttonStyle(.borderedProminent)
                    .tint(EigoLensTheme.primary)
                    .disabled(key.wrappedValue.isEmpty)

                if !key.wrappedValue.isEmpty {
                    Button(action: {
                        key.wrappedValue = ""
                        onSave()
                    }) {
                        Image(systemName: "trash")
                            .foregroundStyle(EigoLensTheme.error)
                    }
                }
            }
        }
        .padding(.vertical, 4)
    }

    private func infoRow(label: String, value: String) -> some View {
        HStack {
            Text(label)
            Spacer()
            Text(value)
                .foregroundStyle(.secondary)
        }
    }

    private func tokenUsageCard(providerName: String, inputTokens: Int, outputTokens: Int, pricing: TokenUsageStore.Pricing, onReset: @escaping () -> Void) -> some View {
        let cost = TokenUsageStore.estimatedCost(inputTokens: inputTokens, outputTokens: outputTokens, pricing: pricing)
        let hasUsage = inputTokens > 0 || outputTokens > 0

        return VStack(alignment: .leading, spacing: 8) {
            Text(providerName)
                .font(EigoLensTheme.titleMedium)

            HStack {
                Text("Input")
                    .font(EigoLensTheme.bodySmall)
                    .foregroundStyle(.secondary)
                Spacer()
                Text(inputTokens.formatted())
                    .font(EigoLensTheme.bodyMedium)
                    .monospacedDigit()
            }
            HStack {
                Text("Output")
                    .font(EigoLensTheme.bodySmall)
                    .foregroundStyle(.secondary)
                Spacer()
                Text(outputTokens.formatted())
                    .font(EigoLensTheme.bodyMedium)
                    .monospacedDigit()
            }
            HStack {
                Text("Est. cost")
                    .font(EigoLensTheme.bodySmall)
                    .foregroundStyle(.secondary)
                Spacer()
                Text(String(format: "$%.4f", cost))
                    .font(EigoLensTheme.bodyMedium)
                    .monospacedDigit()
            }

            if hasUsage {
                Button("Reset", role: .destructive, action: onReset)
                    .font(EigoLensTheme.bodySmall)
            }
        }
        .padding(.vertical, 4)
    }

    private func featureRow(icon: String, title: String, description: String) -> some View {
        HStack(spacing: 12) {
            Image(systemName: icon)
                .font(.title3)
                .foregroundStyle(EigoLensTheme.primary)
                .frame(width: 32)
            VStack(alignment: .leading) {
                Text(title)
                    .font(EigoLensTheme.bodyLarge)
                Text(description)
                    .font(EigoLensTheme.bodySmall)
                    .foregroundStyle(.secondary)
            }
        }
    }
}

File: app/src/main/java/com/jworks/eigolens/ui/settings/SettingsScreen.kt
package com.jworks.eigolens.ui.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.TextButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Slider
import androidx.compose.material3.SliderDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.unit.dp
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp
import java.text.NumberFormat
import java.util.Locale
import androidx.hilt.navigation.compose.hiltViewModel
import com.jworks.eigolens.R

@Composable
fun SettingsScreen(
    onBackClick: () -> Unit,
    viewModel: SettingsViewModel = hiltViewModel()
) {
    val claudeKey by viewModel.claudeApiKey.collectAsState()
    val geminiKey by viewModel.geminiApiKey.collectAsState()
    val activeProvider by viewModel.activeProvider.collectAsState()
    val availableProviders by viewModel.availableProviders.collectAsState()
    val allProviderNames by viewModel.allProviderNames.collectAsState()
    val settings by viewModel.settings.collectAsState()

    var claudeKeyInput by remember(claudeKey) { mutableStateOf(claudeKey) }
    var geminiKeyInput by remember(geminiKey) { mutableStateOf(geminiKey) }

    val providersForUi = if (allProviderNames.isNotEmpty()) allProviderNames else listOf("Claude", "Gemini")
    // Track which provider tab is selected (for showing the right key editor)
    var selectedProviderTab by remember { mutableStateOf(activeProvider ?: providersForUi.firstOrNull() ?: "Claude") }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .verticalScroll(rememberScrollState())
            .padding(16.dp)
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically
        ) {
            IconButton(onClick = onBackClick) {
                Icon(
                    painter = painterResource(id = R.drawable.ic_arrow_back),
                    contentDescription = "Back",
                    tint = Color.White,
                    modifier = Modifier.size(24.dp)
                )
            }
            Text(
                text = "Settings",
                style = MaterialTheme.typography.headlineSmall,
                color = Color.White,
                modifier = Modifier.padding(start = 8.dp)
            )
        }

        Spacer(modifier = Modifier.height(24.dp))

        Text(
            text = "AI Analysis",
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.primary
        )

        Spacer(modifier = Modifier.height(6.dp))

        Text(
            text = "Add API keys to enable phrase/paragraph analysis for circled text.",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        Spacer(modifier = Modifier.height(16.dp))

        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(
                containerColor = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f)
            )
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(12.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Text(
                    text = "Provider",
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                SingleChoiceSegmentedButtonRow(
                    modifier = Modifier.fillMaxWidth()
                ) {
                    providersForUi.forEachIndexed { index, provider ->
                        SegmentedButton(
                            selected = provider == selectedProviderTab,
                            onClick = {
                                selectedProviderTab = provider
                                if (provider in availableProviders) {
                                    viewModel.setActiveProvider(provider)
                                }
                            },
                            shape = SegmentedButtonDefaults.itemShape(index = index, count = providersForUi.size)
                        ) {
                            Text(provider)
                        }
                    }
                }
            }
        }

        Spacer(modifier = Modifier.height(16.dp))

        // Show key editor for the selected provider only
        when (selectedProviderTab) {
            "Claude" -> KeyEditorCard(
                title = "Claude API Key",
                placeholder = "sk-ant-...",
                input = claudeKeyInput,
                saved = claudeKey,
                onInputChanged = { claudeKeyInput = it },
                onSave = { viewModel.setClaudeApiKey(claudeKeyInput.trim()) },
                onClear = { claudeKeyInput = "" }
            )
            "Gemini" -> KeyEditorCard(
                title = "Gemini API Key",
                placeholder = "AIza...",
                input = geminiKeyInput,
                saved = geminiKey,
                onInputChanged = { geminiKeyInput = it },
                onSave = { viewModel.setGeminiApiKey(geminiKeyInput.trim()) },
                onClear = { geminiKeyInput = "" }
            )
        }

        Spacer(modifier = Modifier.height(24.dp))
        HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
        Spacer(modifier = Modifier.height(24.dp))

        Text(
            text = "API Usage",
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.primary
        )

        Spacer(modifier = Modifier.height(6.dp))

        Text(
            text = "Cumulative token usage and estimated cost per provider.",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )

        Spacer(modifier = Modifier.height(12.dp))

        TokenUsageCard(
            providerName = "Claude Haiku",
            inputTokens = settings.claudeInputTokens,
            outputTokens = settings.claudeOutputTokens,
            inputRatePerMillion = 0.80,
            outputRatePerMillion = 4.00,
            onReset = { viewModel.resetTokenUsage("Claude") }
        )

        Spacer(modifier = Modifier.height(8.dp))

        TokenUsageCard(
            providerName = "Gemini 2.5 Flash",
            inputTokens = settings.geminiInputTokens,
            outputTokens = settings.geminiOutputTokens,
            inputRatePerMillion = 0.15,
            outputRatePerMillion = 0.60,
            onReset = { viewModel.resetTokenUsage("Gemini") }
        )

        Spacer(modifier = Modifier.height(24.dp))
        HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
        Spacer(modifier = Modifier.height(24.dp))

        Text(
            text = "Display",
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.primary
        )

        Spacer(modifier = Modifier.height(12.dp))

        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(
                containerColor = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f)
            )
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(12.dp)
            ) {
                Text(
                    text = "IPA Font Size",
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Spacer(modifier = Modifier.height(4.dp))
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = "Small",
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                    Slider(
                        value = settings.ipaFontScale,
                        onValueChange = { viewModel.setIpaFontScale(it) },
                        valueRange = 0.3f..1.0f,
                        colors = SliderDefaults.colors(
                            thumbColor = MaterialTheme.colorScheme.primary,
                            activeTrackColor = MaterialTheme.colorScheme.primary
                        ),
                        modifier = Modifier
                            .weight(1f)
                            .padding(horizontal = 8.dp)
                    )
                    Text(
                        text = "Large",
                        style = MaterialTheme.typography.labelSmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
                Text(
                    text = "Preview: /ˈsæm.pəl/",
                    style = MaterialTheme.typography.bodySmall,
                    fontStyle = androidx.compose.ui.text.font.FontStyle.Italic,
                    color = Color(0xFF00E6FF),
                    fontSize = (6 + (settings.ipaFontScale * 8)).sp
                )
            }
        }

        Spacer(modifier = Modifier.height(24.dp))
        HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
        Spacer(modifier = Modifier.height(24.dp))

        Text(
            text = "About",
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.primary
        )

        Spacer(modifier = Modifier.height(12.dp))

        SettingsItem(label = "Version", value = "0.4.0")
        SettingsItem(label = "Dictionary", value = "WordNet (147,000+ words)")
        SettingsItem(label = "OCR Engine", value = "ML Kit Text Recognition")

        Spacer(modifier = Modifier.height(24.dp))
        HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
        Spacer(modifier = Modifier.height(24.dp))

        Text(
            text = "Features",
            style = MaterialTheme.typography.titleMedium,
            color = MaterialTheme.colorScheme.primary
        )

        Spacer(modifier = Modifier.height(12.dp))

        SettingsItem(label = "Readability Analysis", value = "Flesch-Kincaid, SMOG, Coleman-Liau")
        SettingsItem(label = "NLP", value = "POS Tagging, NER, Lemmatization")
        SettingsItem(label = "Export", value = "PDF (A4 format)")

        Spacer(modifier = Modifier.height(32.dp))

        Text(
            text = "EigoLens by JWorks",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(bottom = 16.dp)
        )
    }
}

@Composable
private fun KeyEditorCard(
    title: String,
    placeholder: String,
    input: String,
    saved: String,
    onInputChanged: (String) -> Unit,
    onSave: () -> Unit,
    onClear: () -> Unit
) {
    val dirty = input != saved

    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f)
        )
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(12.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            Text(
                text = title,
                style = MaterialTheme.typography.titleSmall,
                color = MaterialTheme.colorScheme.onSurface
            )

            OutlinedTextField(
                value = input,
                onValueChange = onInputChanged,
                placeholder = { Text(placeholder) },
                visualTransformation = PasswordVisualTransformation(),
                singleLine = true,
                modifier = Modifier.fillMaxWidth()
            )

            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                Button(
                    onClick = onSave,
                    enabled = dirty && input.isNotBlank()
                ) {
                    Text("Save")
                }

                OutlinedButton(
                    onClick = onClear,
                    enabled = input.isNotEmpty()
                ) {
                    Text("Clear")
                }

                Spacer(modifier = Modifier.weight(1f))

                Text(
                    text = if (saved.isNotBlank()) "Saved" else "Not saved",
                    style = MaterialTheme.typography.labelMedium,
                    color = if (saved.isNotBlank()) {
                        MaterialTheme.colorScheme.primary
                    } else {
                        MaterialTheme.colorScheme.onSurfaceVariant
                    }
                )
            }
        }
    }
}

@Composable
private fun TokenUsageCard(
    providerName: String,
    inputTokens: Long,
    outputTokens: Long,
    inputRatePerMillion: Double,
    outputRatePerMillion: Double,
    onReset: () -> Unit
) {
    val fmt = remember { NumberFormat.getNumberInstance(Locale.US) }
    val cost = (inputTokens * inputRatePerMillion + outputTokens * outputRatePerMillion) / 1_000_000.0
    val hasUsage = inputTokens > 0 || outputTokens > 0

    Card(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = MaterialTheme.colorScheme.surfaceVariant.copy(alpha = 0.4f)
        )
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(12.dp),
            verticalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = providerName,
                    style = MaterialTheme.typography.titleSmall,
                    color = MaterialTheme.colorScheme.onSurface
                )
                if (hasUsage) {
                    TextButton(onClick = onReset) {
                        Text("Reset", style = MaterialTheme.typography.labelSmall)
                    }
                }
            }

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Text(
                    text = "Input tokens",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Text(
                    text = fmt.format(inputTokens),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurface
                )
            }

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Text(
                    text = "Output tokens",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Text(
                    text = fmt.format(outputTokens),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurface
                )
            }

            HorizontalDivider(
                modifier = Modifier.padding(vertical = 4.dp),
                color = MaterialTheme.colorScheme.outlineVariant.copy(alpha = 0.5f)
            )

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Text(
                    text = "Estimated cost",
                    style = MaterialTheme.typography.bodySmall,
                    fontWeight = FontWeight.Medium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                Text(
                    text = "$${String.format(Locale.US, "%.4f", cost)}",
                    style = MaterialTheme.typography.bodySmall,
                    fontWeight = FontWeight.Medium,
                    color = MaterialTheme.colorScheme.primary
                )
            }
        }
    }
}

@Composable
private fun SettingsItem(label: String, value: String) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyLarge,
            modifier = Modifier.weight(1f)
        )
        Text(
            text = value,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
    }
}
